File: first_and_last_position.py
# O(LOG(N)) TIME AND O(1) SPACE WHERE N IS LEN(NUMS)
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        start = -1
        end = -1
        start = self.startBinary(nums,target)
        if start != -1:
            end = self.endBinary(nums,target)
        return [start,end]
    
    def startBinary(self,nums,target):
        left = 0
        right = len(nums)-1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                if mid == 0 or nums[mid-1] != target:
                    return mid
                else:
                    right = mid - 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return -1
    
    def endBinary(self,nums,target):
        left = 0
        right = len(nums)-1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                if mid == len(nums)-1 or nums[mid+1] != target:
                    return mid
                else:
                    left = mid + 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return -1
    

File: FirstAndLastPosition.cpp
//Time Complexity - O(log(n))
//Space Complexity - O(1)
//Problem ran on leet code successfully
//No problems faced

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int noEle=nums.size();
        
        int low=0;
        int high=noEle-1;
        
        int leftMostIndex=-1;
        int rightMostIndex=-1;
        while(low<=high)
        {
            int middleElement=low+ (high-low)/2;
            
            if(target<=nums[middleElement])
            {
                if(nums[middleElement]==target)
                {
                    leftMostIndex=middleElement;
                    if(middleElement==0 || nums[middleElement-1]!=target)
                    {
                        break;
                    }
                    
                }
               high=middleElement-1; 
            }
            else
            {
                low=middleElement+1;
            }
        }
        
        
        
        low=0;
        high=noEle-1;
        while(low<=high)
        {
            int middleElement=low+ (high-low)/2;
            
            if(target>=nums[middleElement])
            {
                if(nums[middleElement]==target)
                {
                    rightMostIndex=middleElement;
                    if(middleElement==noEle-1 || nums[middleElement+1]!=target)
                    {
                        break;
                    }
                }
               low=middleElement+1; 
            }
            else
            {
                high=middleElement-1;
            }
        }
        vector<int> returnVector;
        returnVector.push_back(leftMostIndex);
        returnVector.push_back(rightMostIndex);
        
        return returnVector;
    }
};